src/primitives/triangle/mod.rs

```diff
@@ -180,11 +184,122 @@ impl Triangle {
+        }
+
+        let line_a = Line::new(p1, p2);
+        let line_b = Line::new(p1, p3);
+        let line_c = Line::new(p2, p3);
+
+        let first = bresenham_scanline_intersection(&line_b, scanline_y)?;
+        let second = bresenham_scanline_intersection(&line_a, scanline_y)
+            .or_else(|| bresenham_scanline_intersection(&line_c, scanline_y))?;
+
+        Some(Line::new(
+            first.start.component_min(second.start),
+            first.end.component_max(second.end),
+        ))
+    }
+
+    /// Generate a line join for each corner of the triangle.
+    fn joins(&self, stroke_width: u32, stroke_offset: StrokeOffset) -> [LineJoin; 3] {
+        [
+            LineJoin::from_points(self.p3, self.p1, self.p2, stroke_width, stroke_offset),
+            LineJoin::from_points(self.p1, self.p2, self.p3, stroke_width, stroke_offset),
+            LineJoin::from_points(self.p2, self.p3, self.p1, stroke_width, stroke_offset),
+        ]
+    }
+
+    /// Compute whether a triangle with thick stroke has a hole in its center or is completely
+    /// filled by stroke.
+    // PERF: This doesn't need to compute the entire join, much like how `thick_stroke_inset`
+    // doesn't
+    pub(in crate::primitives::triangle) fn is_collapsed(
+        &self,
+        stroke_width: u32,
+        stroke_offset: StrokeOffset,
+    ) -> bool {
+        let joins = self.joins(stroke_width, stroke_offset);
+
+        joins.iter().enumerate().any(|(i, join)| {
+            // Quick check: if the join is degenerate, no hole can occur.
+            if join.is_degenerate() {
+                return true;
+            }
+
+            // Compute inner-most points of each join. The triangle is sorted clockwise, so that's
+            // the right-side point. The `first_edge_end` and `second_edge_start` points are always
+            // the same in this case, as this is the "pinched" side of the join, so we'll
+            // arbitrarily pick `first_edge_end`.
+            let inner_point = join.first_edge_end.right;
+
+            // Find opposite edge to the given point.
+            let opposite = {
+                let start = self[(i + 1) % 3];
+                let end = self[(i + 2) % 3];
+
+                // Get right side extent (triangle is sorted clockwise, remember)
+                Line::new(start, end).extents(stroke_width, stroke_offset).1
+            };
+
+            // If the inner point is to the left of the opposite side line, the triangle edges self-
+            // intersect, so the triangle is collapsed.
+            opposite.side(inner_point) >= 0
+        })
+    }
+}
+
+impl core::ops::Index<usize> for Triangle {
+    type Output = Point;
+
+    fn index(&self, idx: usize) -> &Self::Output {
+        match idx {
+            0 => &self.p1,
+            1 => &self.p2,
+            2 => &self.p3,
+            _ => unreachable!(),
+        }
+    }
```

Comment:
I would rather not change the `Triangle` structure as it's public. This commend did make me realise that the `Index` impl is public, so I've changed that to a more private `.vertex` method on `Triangle`. It still has an `unreachable()` in it, but this time it really should be as the index wrapping is done in the method.

Reply:
OK, that's better. I still think that it might be beneficial to use a points array for `Triangle`, but that's outside the scope of this PR. Or did we already discuss this in the past? I'm not sure, but I vaguely remember something.

Reply:
I vaguely recall some discussions around this. I wouldn't be opposed to throwing another braking change on the pile, but I'm not sure what benefits the change would bring beyond easier indexing.

Reply:
IMO it would be more consistent with `Polyline` and (in the future) `Polygon`. They use a slice instead of an array, but the syntax is the same in many cases.

Reply:
That's a very good point. I think it would also allow (hopefully) quite a lot of code reuse in the scanline iterators too. I'll open an issue.